#+title: Imputation and Admixture for lcWGS in one goal
[[https://github.com/Zilong-Li/phaseless/actions/workflows/linux.yml/badge.svg]]
[[https://github.com/Zilong-Li/phaseless/actions/workflows/mac.yml/badge.svg]]

Phaseless is designed for genotype imputation and admixture inference using low coverage sequencing data.
Firstly, the imputation model is in the spirit of [[https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1424677/][fastPHASE]] model but with genotype likelihood as input, and likewise [[https://www.nature.com/articles/ng.3594][STITCH]] works on raw read input. Next, the admixture inference is modeled on the haplotype cluster information from the fastphase model.

* Build

Assume you have =htslib= installed in your system path.

#+begin_src shell
git clone https://github.com/Zilong-Li/phaseless
cd phaseless/src && make -j4
#+end_src

If you follow [[https://github.com/samtools/htslib][htslib]] installation guide and have it in your customized path
#+begin_src shell
cd phaseless/src && make HTSLIB=/path/to/your/htslib -j4
#+end_src

* Usage
- run admixture with bealge genotype likelihoods as input
#+begin_src shell
phaseless -g data/bgl.gz -o outdir -c 10 -k 3 -n 20
#+end_src
- run admixture for different =K= with same binary file as input
#+begin_src shell
phaseless -b outdir/pars.bin -k 9 -n 40
#+end_src
- run imputation/phasing only
#+begin_src shell
phaseless -g data/bgl.gz -o outdir -c 10 -n 6 -P 1
#+end_src

* Changes
check out the [[file:news.org][news]] file.
